### PR TITLE
[BPF] Fix tracking packets that host forwards to devices not managed by calico

### PR DESCRIPTION
**Pull request by StevenTigera**
## Description

Adds an option to list devices from which we always let conntrack to track traffic so that if there are interfaces like docker0 and docker injects rules to forward traffic that goes to the host to docker, we do not discrupt NAT on returning traffic.

fixes https://github.com/projectcalico/calico/issues/7420

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
BPF mode: add felix configuration parameter to allow NOTRACK to be disabled for certain external interfaces interfaces.  Useful for workaround conflicts with non-Calico networked containers etc.
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

File: api/pkg/apis/projectcalico/v3/felixconfig.go [hunk 1]
```diff
@@ -442,6 +442,10 @@ type FelixConfigurationSpec struct {
 	// BPFPolicyDebugEnabled when true, Felix records detailed information
 	// about the BPF policy programs, which can be examined with the calico-bpf command-line tool.
 	BPFPolicyDebugEnabled *bool `json:"bpfPolicyDebugEnabled,omitempty"`
+	// BPFForceTrackPacketsFromIfaces forces BPF to track packets from a list of specified interfaces.
+	// Essentially this excludes the specified list of interfaces from being NOTRACKed in IP Tables.
+	BPFForceTrackPacketsFromIfaces *[]string `json:"bpfForceTrackPacketsFromIfaces,omitempty"`
```
Comment (fasaxc) [inline api/pkg/apis/projectcalico/v3/felixconfig.go:447]: Think this field needs validation in libcalico-go to make sure that it is a good list of interface patterns.  That validation probably needs to be duplicated in felix config parsing.  I think we already have a regex for validating interface patterns.

File: felix/config/config_params.go [hunk 2]
```diff
@@ -962,6 +963,8 @@ func loadParams() {
 			}
 		case "cidr-list":
 			param = &CIDRListParam{}
+		case "string-slice":
+			param = &StringSliceParam{}
```
Comment (fasaxc) [inline felix/config/config_params.go:967]: As noted above, I think you need validation that the string values are valid interface patterns.  Perhaps add a validation regex to your new  `StringSliceParam` ?

File: felix/rules/static.go [hunk 3]
```diff
@@ -1144,40 +1144,53 @@ func (r *DefaultRuleRenderer) StaticBPFModeRawChains(ipVersion uint8,
 		Rules: rawRules,
 	}
 
-	bpfUntrackedFlowChain := &Chain{
-		Name:  ChainRawUntrackedFlows,
-		Rules: []Rule{},
-	}
+	// BPF Untracked Flow Rules:
+	var bpfUntrackedFlowRules []Rule
+	var bpfUntrackedFlowChain *Chain
```
Comment (fasaxc) [inline felix/rules/static.go:1149]: Minor: as a general rule, it's best to minimise scope of variables and I don't think bpfUntrackedFlowChain is used until line 1191 so may as well declare it there.

File: api/pkg/apis/projectcalico/v3/felixconfig.go [hunk 4]
```diff
@@ -442,6 +442,10 @@ type FelixConfigurationSpec struct {
 	// BPFPolicyDebugEnabled when true, Felix records detailed information
 	// about the BPF policy programs, which can be examined with the calico-bpf command-line tool.
 	BPFPolicyDebugEnabled *bool `json:"bpfPolicyDebugEnabled,omitempty"`
+	// BPFForceTrackPacketsFromIfaces forces BPF to track packets from a list of specified interfaces.
+	// Essentially this excludes the specified list of interfaces from being NOTRACKed in IP Tables.
```
Comment (fasaxc) [inline api/pkg/apis/projectcalico/v3/felixconfig.go:446]:
```suggestion
	// BPFForceTrackPacketsFromIfaces in BPF mode, forces traffic from these interfaces 
	// to skip Calico's iptables NOTRACK rule, allowing traffic from those interfaces to be
	// tracked by Linux conntrack.  Should only be used for interfaces that are not used for
	// the Calico fabric.  For example, a docker bridge device for non-Calico-networked 
	// containers.
```
Suggest rewording to avoid implying that BPF is doing the tracking

File: felix/config/config_params.go [hunk 5]
```diff
@@ -196,6 +196,7 @@ type Config struct {
 	BPFHostConntrackBypass             bool             `config:"bool;true"`
 	BPFEnforceRPF                      string           `config:"oneof(Disabled,Strict,Loose);Loose;non-zero"`
 	BPFPolicyDebugEnabled              bool             `config:"bool;true"`
+	BPFForceTrackPacketsFromIfaces     []string         `config:"string-slice;;die-on-fail"`
```
Comment (fasaxc) [inline felix/config/config_params.go:199]: Should we make the default useful, perhaps `docker+`?  Do we need `die-on-fail`; think the default behaviour might be good enough for this param.

File: felix/rules/static.go [hunk 6]
```diff
@@ -1144,40 +1144,53 @@ func (r *DefaultRuleRenderer) StaticBPFModeRawChains(ipVersion uint8,
...
-					Action:  ReturnAction{},
-					Comment: []string{"MarkSeenNATOutgoing Mark"},
-				},
-				Rule{
-					Action: NoTrackAction{},
-				},
-			},
+		// Iterate all BPF interfaces forced to track packets and append rule.
+		for _, interfaceName := range r.BPFForceTrackPacketsFromIfaces {
+			if len(interfaceName) > 0 {
```
Comment (fasaxc) [inline felix/rules/static.go:1154]: Should add a UT for the new feature

File: api/pkg/apis/projectcalico/v3/felixconfig.go (hunk 1 above)
Comment (StevenTigera) [reply api/pkg/apis/projectcalico/v3/felixconfig.go:447]: Implemented as recommended; found ifaceFilterRegex which is preferred over interfaceRegex as this accommodates wildcards e.g. "docker+"

File: felix/config/config_params.go (hunk 5 above)
Comment (StevenTigera) [reply felix/config/config_params.go:199]: Implemented as suggested: make `docker+` the default and remove `die-on-fail`

File: felix/config/config_params.go (hunk 2 above)
Comment (StevenTigera) [reply felix/config/config_params.go:967]: Could I pls choose to not implement this suggestion because `StringSliceParam` was named added as a generic helper function to parse the raw input string and return as string slice i.e. no affinity to any input like valid interface(s) thus not subject to validation - the validation has been implemented above in the Felix config however which, if input there failed then, would not be need to be re-validated when set as config_params - thanks

File: api/pkg/apis/projectcalico/v3/felixconfig.go (hunk 4 above)
Comment (tomastigera) [reply api/pkg/apis/projectcalico/v3/felixconfig.go:446]: Also it seems like we allow iptables style prefix like `docker+` so that needs to be documented here, saying, that you can use a specific device like `docker0` or a prefix like `docker+`. Note that we use regexps in most places. It should also state what the default is like `[Default: docker+]`, see the other options for examples.

File: felix/rules/static.go (hunk 3 above)
Comment (StevenTigera) [reply felix/rules/static.go:1149]: Updated as recommended

File: felix/rules/static.go (hunk 6 above)
Comment (StevenTigera) [reply felix/rules/static.go:1154]: Additional unit tests for BPF mode raw chains

File: libcalico-go/config/crd/crd.projectcalico.org_felixconfigurations.yaml [hunk 7]
```diff
@@ -108,6 +108,13 @@ spec:
                   node appears to use the IP of the ingress node; this requires a
                   permissive L2 network.  [Default: Tunnel]'
                 type: string
+              bpfForceTrackPacketsFromIfaces:
+                description: BPFForceTrackPacketsFromIfaces forces BPF to track packets
```
[the remark this answers is not in the document]
Comment (StevenTigera) [reply libcalico-go/config/crd/crd.projectcalico.org_felixconfigurations.yaml]: Thanks @tomastigera - you checked this too soon :)  I wanted to address all comments and regression test before git commit / push

File: api/pkg/apis/projectcalico/v3/felixconfig.go (hunk 4 above)
Comment (StevenTigera) [reply api/pkg/apis/projectcalico/v3/felixconfig.go]: Thanks @tomastigera - I realized that `[Default: docker+]` was omitted - added now and re did `make generate` and confirm visible upon `kubectl explain felixconfig.spec | grep -7 bpfForceTrackPacketsFromIfaces`

File: felix/config/config_params.go (hunk 2 above)
Comment (fasaxc) [reply felix/config/config_params.go:967]: I do think you need validation here because Felix also supports env-var based configuration in manifest mode (and a config file, which is used in OpenStack).

I think the change here would be something like this:
```
case "interface-name-slice":
	param = &StringSliceParam{ValidationRegex: ...}
```
So your StringSliceParam could stay generic, but you'd have a particular sub-type with particular validation rules.

Comment (StevenTigera) [reply felix/config/config_params.go:967]: Implemented string slice parameter validation as recommended

File: felix/config/config_params_test.go [hunk 8]
```diff
@@ -504,6 +504,11 @@ var _ = DescribeTable("Config parsing",
 		map[string]time.Duration{"a": 0}, false),
 	// Not a required parameter so a bad value is translated to nil:
 	Entry("HealthTimeoutOverrides non-duration", "HealthTimeoutOverrides", "foo=bar", map[string]time.Duration(nil), false),
+
+	Entry("BPFForceTrackPacketsFromIfaces", "BPFForceTrackPacketsFromIfaces", "", []string{"docker+"}),
+	Entry("BPFForceTrackPacketsFromIfaces", "BPFForceTrackPacketsFromIfaces", "docker0", []string{"docker0"}),
+	Entry("BPFForceTrackPacketsFromIfaces", "BPFForceTrackPacketsFromIfaces", "docker+", []string{"docker+"}),
+	Entry("BPFForceTrackPacketsFromIfaces", "BPFForceTrackPacketsFromIfaces", "docker0,docker1", []string{"docker0", "docker1"}),
```
Comment (fasaxc) [inline felix/config/config_params_test.go:511]: Please add an entry that fails validation.

**Review by fasaxc — CHANGES_REQUESTED** [review]
One tiny request on the test side, otherwise good to go

File: felix/config/config_params_test.go (hunk 8 above)
Comment (StevenTigera) [reply felix/config/config_params_test.go:511]: Thanks @fasaxc I as per request I am trying to add an entry that fails validation something like
```Entry("BPFForceTrackPacketsFromIfaces", "BPFForceTrackPacketsFromIfaces", "cali@123", []string{"docker+"}, true),```
But in its current form the test will not pass because the config params `die-on-failure` was removed from `BPFForceTrackPacketsFromIfaces` as suggested prior.  I see when this is the case the expected Error will never be true as the err is reverted to nil and the updated parameter will revert to its default value i.e. `docker+` in this case.  Therefore, I can add the following test instead
```Entry("BPFForceTrackPacketsFromIfaces", "BPFForceTrackPacketsFromIfaces", "cali@123", []string{"docker+"}),```
 i.e. expected Error false but pls let me know if you would like the config params `die-on-failure` validation to be re-instated - thanks

Comment (fasaxc) [reply felix/config/config_params_test.go:511]: SGTM

Comment (StevenTigera) [reply felix/config/config_params_test.go:511]: Align config params test with params type test and include valid and invalid use cases

**Review by fasaxc — APPROVED** [review]
LGTM, probably worth squashing on merge?

**Comment by StevenTigera** [comment]
> LGTM, probably worth squashing on merge?

Done
